partner/WHMCS/modules/servers/jetpack/jetpack.php

```diff
@@ -0,0 +1,365 @@
+    } else {
+        $errors = get_cancellation_errors_from_response($response);
+        return $errors;
+    }
+
+}
+
+/**
+ * Get a Jetpack partner access token using the client_id and client secret
+ * stored when the product was created in the WHMCS product settings. If the
+ * response does not explicitly contain an access token provisioning or cancellation
+ * cannot be attempted so return an error string.
+ *
+ *
+ * @param array $params WHMCS params
+ * @return mixed A string with the access token or an error string beginning with 'JETPACK MODULE' indicating that
+ * an access token was not retrieved for provisioning or cancelling.
+ */
+function get_access_token($params)
+{
+
+    $oauth_url = "https://public-api.wordpress.com/oauth2/token";
+
+    $credentials = [
+        'client_id' => $params['configoption1'],
+        'client_secret' => $params['configoption2'],
+        'grant_type' => 'client_credentials',
+        'scope' => 'jetpack-partner'
+    ];
+
+    $response = make_api_request($oauth_url, null, $credentials);
+
+    if (isset($response->access_token)) {
+        return $response->access_token;
+    } else {
+        $errors = get_authentication_errors_from_response($response);
+        return $errors;
+    }
+}
+
+
+/**
+ * Make an API request for authenticating and provisioning or cancelling a Jetpack plan or getting an access token.
+ * Include the http status in the response.
+ *
+ * @param string $url where to make the request to
+ * @param null $auth access token to make a provisioning or cancellation request
+ * @param null $data form data for the api request
+ * @return mixed json_decoded response
+ * @throws Exception On a curl error or an empty body an Exception will be thrown.
+ */
+function make_api_request($url, $auth = null, $data = null)
+{
+
+    if (isset($auth)) {
+        $auth = "Authorization: Bearer " . $auth;
+    }
+
+    $curl = curl_init();
+    curl_setopt_array($curl, [
+        CURLOPT_HTTPHEADER => [$auth],
+        CURLOPT_URL => $url,
+        CURLOPT_RETURNTRANSFER => true,
+        CURLOPT_ENCODING => "",
+        CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
+        CURLOPT_POSTFIELDS => $data,
+        CURLOPT_CUSTOMREQUEST => "POST"
+    ]);
+
+    $response = curl_exec($curl);
+    $http_code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+    if (curl_error($curl)) {
+        throw new Exception('Unable to connect: ' . curl_errno($curl) . ' - ' . curl_error($curl));
+    } elseif (empty($response)) {
+        throw new Exception('Empty response');
+    }
+
+
+    $decoded_response = json_decode($response);
+    $decoded_response->http_status = $http_code;
+    curl_close($curl);
+
+    return $decoded_response;
+}
+
+/**
+ * Save the next_url for Jetpack activation/setup to the order for the client. If a next url is not
+ * and the pending param is true save a message letting the user know a plan is pending and will
```

Comment:
"If a next url is not and" should probably be something like, "If a next url is not present or is empty and"

Reply:
Also fixed.